### ISSUE TITLE
PNG exporting does not work because it assumes chart can be reinitialized

### ISSUE THREAD

**Issue by mitar** [highcharts/highcharts#4486]
See [reproduction here](http://jsfiddle.net/xtahp2mk/). Open JSFiddle and then try exporting a PNG image.

**Comment by pawelfus** [comment]
Isn't that problem with asyns.js? 

Workaround (solution?): http://jsfiddle.net/fbxdzukw/

**Comment by mitar** [comment]
Good workaround. I think the issue is that there is this hidden assumption you can just initialize it as many times as you want. I have not seen this documented anywhere.

**Comment by pawelfus** [comment]
Yes, and I think this is proper. For example, user can add extra drawing to the chart in callback using `Renderer`. It would be a bug if exported chart won't include those drawings, right? 

It looks like callback can be called just once, for some reasons?

**Comment by mitar** [comment]
> It looks like callback can be called just once, for some reasons?

Yes. Because it is the same value. It is not recreated.

**Comment by mitar** [comment]
> Yes, and I think this is proper. For example, user can add extra drawing to the chart in callback using Renderer. It would be a bug if exported chart won't include those drawings, right? 

True, but on the other hand there might be also code which expects to be called only once.

**Comment by pawelfus** [comment]
Ok, but how would we determine how many times callback should be called? Notice, limitation comes from an another library, so I think user should take care of such callback. No errors given with another callback: http://jsfiddle.net/fbxdzukw/1/ 

Even simple flag `true`/`false` would resolve this problem (to call it just once).

**Comment by mitar** [comment]
The issue is that it is unclear what is behavior here. For example, in my case, I have the following order of steps:
* using async, initialize multiple charts
* once all charts are initialized (they reserve the space on the page, which should be quick), start loading data for them
* once all data is loaded, start rendering the charts

Now, if you just call the first call to highcharts again, then things will not be rendered correctly. Because even if I do not call the callback (as in workaround), the real data will not even be added to the chart).

Compare the screenshot:

<img width="1156" alt="untitled2" src="https://cloud.githubusercontent.com/assets/585279/9352215/4e119f4c-4614-11e5-909f-be06c676cfe3.png">

With export:

![chart](https://cloud.githubusercontent.com/assets/585279/9352220/5a3735c0-4614-11e5-92a3-174bd5e09e3c.png)

Navigator is not there. Dates in the right top corner are not there. And also it is much too narrower.

**Comment by pawelfus** [comment]
Please, setup such demo, I will check this. Regarding top right input and narrow chart, please read [this](http://www.highcharts.com/docs/export-module/export-module-overview) article..

**Comment by mitar** [comment]
> Please, setup such demo, I will check this.

Huh, this is a bit tricky. I see that I will anyway have to do some special processing when exporting. But it is not that it is a completely independent new chart created? Because now I added that series are added manually again, in the callback, and I have now series duplicated.

![chart 1](https://cloud.githubusercontent.com/assets/585279/9352458/3d76474e-4616-11e5-91cb-095b6d9b480d.png)

**Comment by pawelfus** [comment]
Yeah, that will duplicate series. Consider these examples:

- http://jsfiddle.net/3gLhpkcf/ (add by button)
- http://jsfiddle.net/3gLhpkcf/1/ (added by callback)

Both examples use `addSeries`. Now, first one adds series, and when exporting is used, we get two series. In other words, chart **stores** information about new series and use them in exporting too. Now use the same logic for the second chart: chart renders with one series and calls callback, so we get two series on the chart (stored). Now exporting is called, which means adding new series in callback, so we get three series on chart (even if only one is visible, because lines overlap). 

>  >  Please, setup such demo, I will check this.

> Huh, this is a bit tricky. 

Not a problem, we can wait. Also, you can use some random data instead or real ones.

**Comment by mitar** [comment]
So if I understand correctly, I should not be adding series in the callback because exporting should do that by itself. But the fact that it does not do that for navigator, is a bug and you would like a reproduction for that?

**Comment by mitar** [comment]
> Regarding top right input and narrow chart, please read this article..

Is there are a callback which could be set to dynamically provide width and height, so that I could inspect the width and height of the rendered chart when clicked in the menu?

**Comment by pawelfus** [comment]
In general, navigator supports only one series at time. It's fine adding series in callback, but add this only once. See [this](http://highcharts.uservoice.com/forums/55896-general) suggestion. 

> Is there are a callback which could be set to dynamically provide width and height, so that I could inspect the width and height of the rendered chart when clicked in the menu?

See [chartOptions](http://api.highcharts.com/highcharts#exporting.chartOptions). There you can set:

    chartOptions: {
      chart: { 
        events: { 
          load: function() { ... } // resize chart in this callback
        }
      }
    }

**Comment by mitar** [comment]
No, I do not want to resize the chart, I would like to measure the chart and use that when exporting. From exporting documentation it seems it uses chart size only if it has a fixed size set. I would like to measure the size with jQuery myself and provide that values to the exporting.

**Comment by mitar** [comment]
> See this suggestion/enhancement.

This is a bad link?

**Comment by mitar** [comment]
> In general, navigator supports only one series at time.

Oh, I am using it for multiple and it works well. :-) But maybe this is the reason why it does not work when exporting.

**Comment by pawelfus** [comment]
Right, proper [link](http://highcharts.uservoice.com/forums/55896-highcharts-javascript-api/suggestions/2361925-allow-navigator-to-have-multiple-data-series) - sorry!

**Comment by mitar** [comment]
Yea, I am adding multiple [series to navigator without problems](https://github.com/wlanslovenija/django-datastream/blob/master/django_datastream/static/datastream/js/datastream.js#L898).

**Comment by mitar** [comment]
Reproduction: http://jsfiddle.net/ztq52305/

You can see the series in the navigator, but if I export the chart, there is no data in the navigator.

**Comment by mitar** [comment]
OK, this seems to work:

```
function (highcharts) {
            // When exporting, charts are recreated by Highcharts. We do not do anything here.
            if (highcharts.options.chart.forExport) {
                return;
            }
            else {
                self.highcharts = highcharts;

                var $container = $(self.highcharts.container);
                highcharts.options.exporting.sourceWidth = $container.outerWidth();
                highcharts.options.exporting.sourceHeight = $container.outerHeight();

                if (callback) callback();
            }
        }
```

![chart 4](https://cloud.githubusercontent.com/assets/585279/9364498/814135ba-4661-11e5-8994-0ed5e0666e66.png)

It sets the correct output size when exporting now. Now only empty navigator is the issue.

Also, I see that y-axis changes the scale when I export. Like that it would refresh somehow its extremes. But this is probably some other issue.

**Comment by mitar** [comment]
So this ticket is now only to report that in exported chart the navigator is empty. See [reproduction](http://jsfiddle.net/ztq52305/). Should I open another ticket instead?

**Comment by pawelfus** [comment]
As I said before, right now multiple series are not supported by navigator, only workaround is to add them after chart is rendered. Exactly the same issues are described [here](http://highcharts.uservoice.com/forums/55896-highcharts-javascript-api/suggestions/2361925-allow-navigator-to-have-multiple-data-series) - I suggest to vote for that feature in first place ;)

**Comment by pawelfus** [comment]
@mitar - multiple series in the navigator are now supported. Did you get a chance to check [`series.showInNavigator`](http://api.highcharts.com/highstock/plotOptions.series.showInNavigator) option?

**Comment by mitar** [comment]
Cool. I will take a look.
